src/liballoc/arc.rs

```diff
@@ -529,6 +531,153 @@ impl<T: ?Sized> Arc<T> {
     }
 }
 
+impl<T: ?Sized> Arc<T> {
+    // Allocates an `ArcInner<T>` with sufficient space for an unsized value
+    unsafe fn allocate_for_ptr(ptr: *const T) -> *mut ArcInner<T> {
+        // Create a fake ArcInner to find allocation size and alignment
+        let fake_ptr = ptr as *mut ArcInner<T>;
+
+        let size = size_of_val(&*fake_ptr);
+        let align = align_of_val(&*fake_ptr);
+
+        let mem = allocate(size, align);
+
+        if mem.is_null() {
+            ::oom();
+        }
+
+        // Initialize the real ArcInner
+        let inner = set_data_ptr(ptr as *mut T, mem) as *mut ArcInner<T>;
+
+        ptr::write(&mut (*inner).strong, atomic::AtomicUsize::new(1));
+        ptr::write(&mut (*inner).weak, atomic::AtomicUsize::new(1));
+
+        inner
+    }
+
+    fn from_box(v: Box<T>) -> Arc<T> {
+        unsafe {
+            let bptr = Box::into_raw(v);
+
+            let value_size = size_of_val(&*bptr);
+            let ptr = Self::allocate_for_ptr(bptr);
+
+            // Copy value as bytes
+            ptr::copy_nonoverlapping(
+                bptr as *const T as *const u8,
+                &mut (*ptr).data as *mut _ as *mut u8,
+                value_size);
+
+            // Free the allocation without dropping its contents
+            box_free(bptr);
+
+            Arc { ptr: Shared::new(ptr) }
+        }
+    }
+}
+
+// Sets the data pointer of a `?Sized` raw pointer.
+//
+// For a slice/trait object, this sets the `data` field and leaves the rest
+// unchanged. For a sized raw pointer, this simply sets the pointer.
+unsafe fn set_data_ptr<T: ?Sized, U>(mut ptr: *mut T, data: *mut U) -> *mut T {
+    ptr::write(&mut ptr as *mut _ as *mut *mut u8, data as *mut u8);
+    ptr
+}
+
+impl<T> Arc<[T]> {
+    fn from_vec(mut v: Vec<T>) -> Arc<[T]> {
+        unsafe {
+            let v_ptr = &v[..] as *const [T];
+            let ptr = Self::allocate_for_ptr(v_ptr);
+
+            ptr::copy_nonoverlapping(
+                v.as_ptr(),
+                &mut (*ptr).data as *mut [T] as *mut T,
+                v.len());
+
+            // Allow the Vec to free its memory, but not destroy its contents
+            v.set_len(0);
+
+            Arc { ptr: Shared::new(ptr) }
+        }
+    }
+}
+
+impl<T: Copy> Arc<[T]> {
+    fn copy_from_slice(v: &[T]) -> Arc<[T]> {
+        unsafe {
+            let v_ptr = v as *const [T];
+            let ptr = Self::allocate_for_ptr(v_ptr);
+
+            ptr::copy_nonoverlapping(
+                v.as_ptr(),
+                &mut (*ptr).data as *mut [T] as *mut T,
+                v.len());
+
+            Arc { ptr: Shared::new(ptr) }
+        }
+    }
+}
+
+impl<T: Clone> Arc<[T]> {
+    fn clone_from_slice(v: &[T]) -> Arc<[T]> {
```

Comment:
This and `copy_from_slice` are only used in one location, right? Could they just be inlined into `From` below?